pkg/services/secrets/kvstore/migrations/migrator.go

```diff
@@ -0,0 +1,51 @@
+package migrations
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	datasources "github.com/grafana/grafana/pkg/services/datasources/service"
+)
+
+var logger = log.New("secret.migration")
+
+// SecretMigrationService is used to migrate legacy secrets to new unified secrets.
+type SecretMigrationService interface {
+	Migrate(ctx context.Context) error
+}
+
+type SecretMigrationServiceImpl struct {
+	Services []SecretMigrationService
+}
+
+func ProvideSecretMigrationService(
+	dataSourceSecretMigrationService *datasources.DataSourceSecretMigrationService,
+) *SecretMigrationServiceImpl {
+	return NewSecretMigrationService(
+		dataSourceSecretMigrationService,
+	)
+}
+
+func NewSecretMigrationService(services ...SecretMigrationService) *SecretMigrationServiceImpl {
+	return &SecretMigrationServiceImpl{services}
+}
+
+// Run migration services. This will block until all services have exited.
+func (s *SecretMigrationServiceImpl) Migrate(ctx context.Context) error {
```

Comment:
Think you might need to consider high availability (HA) setup here as well when there are multiple Grafana instances starting up at the same time. 

Might want to wrap either this in a serverlock or each migration service. Serverlock make sure that only one Grafana instance can run a certain code at a certain time.  See below for example usage:
https://github.com/grafana/grafana/blob/f78b7f1f179e6fb8f2e378e425cd8d0b64762b1b/pkg/services/auth/token_cleanup.go#L15-L22

One drawback with serverlock might be that instances not getting the lock will return an error and fail Grafana to start. It's probably a good thing, since continuing execution after not getting this lock will run provisioning and raise with the secrets migration on another Grafana instance.

Reply:
Will do! I was just discussing that with @leandro-deveikis the other day, we weren't sure how to handle HA setups yet, but that example is enough.